sdkproject/Assets/Mapbox/Core/mapbox-sdk-cs/Platform/Cache/CachingWebFileSource.cs

```diff
@@ -4,55 +4,131 @@
 	using Mapbox.Platform;
 	using System.Collections.Generic;
 	using Mapbox.Unity.Utilities;
+	using Mapbox.Map;
 
-	public class CachingWebFileSource : IFileSource
+	public class CachingWebFileSource : IFileSource, IDisposable
 	{
 
 
+		private bool _disposed;
 		private List<ICache> _caches = new List<ICache>();
 		private string _accessToken;
 
+
 		public CachingWebFileSource(string accessToken)
 		{
 			_accessToken = accessToken;
 		}
 
+
+		#region idisposable
+
+
+		~CachingWebFileSource()
+		{
+			Dispose(false);
+		}
+
+		public void Dispose()
+		{
+			Dispose(true);
+			GC.SuppressFinalize(this);
+		}
+
+		protected virtual void Dispose(bool disposeManagedResources)
+		{
+			if (!_disposed)
+			{
+				if (disposeManagedResources)
+				{
+					for (int i = 0; i < _caches.Count; i++)
+					{
+						IDisposable cache = _caches[i] as IDisposable;
+						if (null != cache)
+						{
+							cache.Dispose();
+							cache = null;
+						}
+					}
+				}
+				_disposed = true;
+			}
+		}
+
+
+		#endregion
+
+
+		/// <summary>
+		/// Add an ICache instance
+		/// </summary>
+		/// <param name="cache">Implementation of ICache</param>
+		/// <returns></returns>
 		public CachingWebFileSource AddCache(ICache cache)
 		{
+			// don't add cache when cache size is 0
+			if (0 == cache.MaxCacheSize)
+			{
+				return this;
+			}
+
 			_caches.Add(cache);
 			return this;
 		}
 
 
-		public IAsyncRequest Request(string uri, Action<Response> callback, int timeout = 10)
+		/// <summary>
+		/// Clear all caches
+		/// </summary>
+		public void Clear()
+		{
+			foreach (var cache in _caches)
+			{
+				cache.Clear();
+			}
+		}
+
+
+		public IAsyncRequest Request(
+			string uri
+			, Action<Response> callback
+			, int timeout = 10
+			, CanonicalTileId tileId = new CanonicalTileId()
+			, string mapId = null
+		)
 		{
 
+			if (string.IsNullOrEmpty(mapId))
```

Comment:
@david-rhodes what would be the benefit of delegating to the cache itself?

This check has to be done anyway and I'd prefer to fail as early as possible. 
I think it wouldn't make sense to fire a request and then throw after it has returned. Time and resources wasted.
Or am I missing something?